dask/dataframe/multi.py

```diff
@@ -1096,7 +1096,7 @@ def stack_partitions(dfs, divisions, join="outer", ignore_order=False, **kwargs)
         # this case we need to pass along the meta object to transform each
         # partition, so they're all equivalent.
         try:
-            df._meta == meta
+            assert df._meta == meta
```

Comment:
Don't think this would have the intended impact of checking for matching meta - with pandas, asserting dataframe equality like this should always raise a `ValueError`:

```
ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

And in cuDF raises an uncaufght `AssertionError` because the result of this `==` operation (an empty dataframe) is falsey.

Is there a utility we can use to assert equality between these meta? `assert_eq` came to mind but imagine this is typically intended for tests only.

Alternatively, if we are only interested in a specific attribute of the metas (columns or index), might make more sense to do an assertion on those.

Reply:
Thanks for reporting @charlesbluca. 

> with pandas, asserting dataframe equality like this should always raise a ValueError:

I think this is the sort of comparison going on here (though if you have a different case, please let me know):

```python
In [1]: import pandas as pd

In [2]: x = pd.DataFrame({"a": []})

In [3]: y = pd.DataFrame({"b": []})

In [4]: x == y
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 x == y

File ~/mambaforge/envs/dask/lib/python3.10/site-packages/pandas/core/ops/common.py:72, in _unpack_zerodim_and_defer.<locals>.new_method(self, other)
     68             return NotImplemented
     70 other = item_from_zerodim(other)
---> 72 return method(self, other)

File ~/mambaforge/envs/dask/lib/python3.10/site-packages/pandas/core/arraylike.py:43, in OpsMixin.__eq__(self, other)
     41 @unpack_zerodim_and_defer("__eq__")
     42 def __eq__(self, other):
---> 43     return self._cmp_method(other, operator.eq)

File ~/mambaforge/envs/dask/lib/python3.10/site-packages/pandas/core/frame.py:7601, in DataFrame._cmp_method(self, other, op)
   7598 def _cmp_method(self, other, op):
   7599     axis = 1  # only relevant for Series other case
-> 7601     self, other = ops.align_method_FRAME(self, other, axis, flex=False, level=None)
   7603     # See GH#4537 for discussion of scalar op behavior
   7604     new_data = self._dispatch_frame_op(other, op, axis=axis)

File ~/mambaforge/envs/dask/lib/python3.10/site-packages/pandas/core/ops/__init__.py:290, in align_method_FRAME(left, right, axis, flex, level)
    288             left, right = left.align(right, join="outer", level=level, copy=False)
    289         else:
--> 290             raise ValueError(
    291                 "Can only compare identically-labeled DataFrame objects"
    292             )
    293 elif isinstance(right, ABCSeries):
    294     # axis=1 is default for DataFrame-with-Series op
    295     axis = left._get_axis_number(axis) if axis is not None else 1

ValueError: Can only compare identically-labeled DataFrame objects
```

> And in cuDF raises an uncaufght AssertionError because the result of this == operation (an empty dataframe) is falsey

Is there a traceback for this somewhere we can look at?

Reply:
Thanks for sharing that case! The one I encountered was part of `dask/dataframe/tests/test_multi.py::test_concat[inner]`; setting a breakpoint right before the assertion:

```
> /home/nfs/charlesb/dev/dask/main/dask/dataframe/multi.py(1100)stack_partitions()
-> assert df._meta == meta
(Pdb) df._meta
Empty DataFrame
Columns: [x, y]
Index: []
(Pdb) meta
Empty DataFrame
Columns: [x, y]
Index: []
(Pdb) assert df._meta == meta
*** ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

Not sure if there's some additional nuance I'm missing, but I would think `match` should be true in a case like this?

> Is there a traceback for this somewhere we can look at?

Yeah, this started cropping up in some query tests running on gpuCI:

https://gpuci.gpuopenanalytics.com/job/dask/job/dask-sql/job/prb/job/dask-sql-prb/1299/CUDA_VER=11.5,LINUX_VER=ubuntu18.04,PYTHON_VER=3.9,RAPIDS_VER=22.10/testReport/junit/tests.unit/test_queries/test_gpu_query_q56_sql_/

Reply:
Thanks for raising this. I did see variants of this `ValueError` for [this assertion](https://github.com/dask/dask/pull/9459/files#diff-ce961485dc8b8afac19632a94751f2cdef3aaa5010cdafcb2966c882b4a5583e) when the test matched a `pd.Series` against a `pd.DataFrame`.

One option could be to cover these types first like this? 

```python
if isinstance(df, pd.DataFrame):
    pd.testing.assert_frame_equal(df._meta, meta)
elif isinstance(df, pd.Series):
    pd.testing.assert_series_equal(df._meta, meta)
else:
    assert df._meta == meta
```

I'm not if this helps with the other points raised here, so if not please ignore it.